mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -0,0 +1,245 @@
+use anyhow::anyhow;
+use async_trait::async_trait;
+use pallas_network::facades::NodeClient;
+use pallas_network::miniprotocols::localstate::queries_v16;
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+
+use crate::chain_observer::interface::*;
+use crate::chain_observer::{ChainAddress, TxDatum};
+use crate::crypto_helper::{KESPeriod, OpCert};
+use crate::entities::StakeDistribution;
+use crate::CardanoNetwork;
+use crate::{entities::Epoch, StdResult};
+
+use super::CardanoCliChainObserver;
+
+/// A runner that uses Pallas library to interact with a Cardano node using N2C Ouroboros mini-protocols
+pub struct PallasChainObserver {
+    socket: PathBuf,
+    network: CardanoNetwork,
+    fallback: Option<super::cli_observer::CardanoCliChainObserver>,
+}
+
+#[derive(Error, Debug)]
+pub enum PallasChainObserverError {
+    #[error("Failed to create new client: {0}")]
+    UnableToCreateClient(anyhow::Error),
+
+    #[error("Acquire failed: {0}")]
+    AcquireFailed(anyhow::Error),
+
+    #[error("Failed to get current era: {0}")]
+    GetCurrentEraFailed(anyhow::Error),
+
+    #[error("Failed to get block epoch number: {0}")]
+    GetBlockEpochNumberFailed(anyhow::Error),
+
+    #[error("Send release failed: {0}")]
+    SendReleaseFailed(anyhow::Error),
+
+    #[error("Send done failed: {0}")]
+    SendDoneFailed(anyhow::Error),
+
+    #[error("Chainsync send done failed: {0}")]
+    ChainSyncSendDoneFailed(anyhow::Error),
+}
```

Comment:
All these errors are coming from a sub-system and they are not tested anywhere. 
This means that you can remove them and pass directly the sub-system errors to the caller with `anyhow` (and use the `with_context` function to create a nice stack trace).